### PR TITLE
feat: step can read trimmed curves and cruves can trim themselves

### PR DESCRIPTION
**Pull request by AxelRinghausen**
* **Please check if the PR fulfills these requirements**
- [x] The commit messages follows our guidelines (explicit, in english)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bug fix:
- [x] New features: Circle3D.trim and BSplineCurve3D.trim
- [ ] Documentation
- [ ] Optimization

* **Other information**:

### CODE REVIEW

File: volmdlr/edges.py [hunk 1]
```diff
@@ -2597,6 +2666,54 @@ def point_at_abscissa(self, curvilinear_abscissa):
         # # Outside of length
         # raise ValueError
 
+    def norm_eval_at_point(self, point: volmdlr.Point3D, resolution=20):
```
Comment (masfaraud) [inline volmdlr/edges.py:2669]: is it the same than normal for other edges? https://github.com/Dessia-tech/volmdlr/blob/1304748eb8d41dd3c1b6220a0fb9d5a7659ef27d/volmdlr/edges.py#L2592

File: volmdlr/step.py [hunk 2]
```diff
@@ -227,35 +229,34 @@ def create_graph(self, draw=False, html=False):
...
-        #         edges=edges,
-        #         options=options)
-        #
-        #     with open('graph_visJS.html', 'wb') as file:
-        #         file.write(s.encode('utf-8'))
-        #
-        #     webbrowser.open('file://' + os.path.realpath('graph_visJS.html'))
+        if html:
+
+            env = Environment(
```
Comment (masfaraud) [inline volmdlr/step.py:234]: Better use template substitution as https://github.com/Dessia-tech/volmdlr/blob/1304748eb8d41dd3c1b6220a0fb9d5a7659ef27d/volmdlr/core.py#L1486

File: volmdlr/step.py [hunk 3]
```diff
@@ -16,6 +16,8 @@
 import plot_data.graph
 
 import webbrowser
+from jinja2 import Environment, PackageLoader, select_autoescape
```
Comment (masfaraud) [inline volmdlr/step.py:19]: Remove jinja2 import, it makes difficult to install volmdlr

**Review by masfaraud — CHANGES_REQUESTED** [review]

File: volmdlr/edges.py (hunk 1 above)
Comment (AxelRinghausen) [reply volmdlr/edges.py:2676]: 'norm_eval_at_point' is short for 'normalized_evaluation_at_point', and search for the value of the normalized evaluation parameter t (between 0 and 1) that would return the given point when the BSplineCurve3D is evaluated at the t value.
Should I change the name of the function for something more explicit ?

File: volmdlr/step.py (hunk 2 above)
Comment (AxelRinghausen) [reply volmdlr/step.py]: Fixed in https://github.com/Dessia-tech/volmdlr/pull/190/commits/3c49b6ae958e2591dd326b81e3fd694d4084c141

File: volmdlr/step.py (hunk 3 above)
Comment (AxelRinghausen) [reply volmdlr/step.py]: Fixed in https://github.com/Dessia-tech/volmdlr/pull/190/commits/3c49b6ae958e2591dd326b81e3fd694d4084c141

File: volmdlr/edges.py (hunk 1 above)
Comment (masfaraud) [reply volmdlr/edges.py:2644]: Can you put it in docstring?

**Review by masfaraud — CHANGES_REQUESTED** [review]

**Review by masfaraud — DISMISSED** [review]

**Review by masfaraud — APPROVED** [review]

**Pull request merged**